### DISCUSSION TITLE
Override function in woocommerce/inc/wp-loop.php

### DISCUSSION THREAD

**Discussion by xoneill007** [bootscore/bootscore#597] [Q&A] [answered]
Thanks in advance,

I'm trying to override function ("custom_loop_product_thumbnail()") in woocommerce/inc/wp-loop.php.

1) Per: "Changing cart and user offcanvas to modal #595", I copied wp-loop.php to the child theme:

`bootscore-child-main/woocommerce/inc/wc-loop.php`

2) Added the following to the child's function.php:

```
function remove_thumbs() {
   require_once('woocommerce/inc/wc-loop.php');
}
add_action('woocommerce_before_shop_loop_item_title', 'remove_thumbs');
```

3) Then modified 'woocommerce/inc/wc-loop.php' with:

`remove_action('woocommerce_before_shop_loop_item_title', 'custom_loop_product_thumbnail', 10);`

It removes all thumbnails from the products page except the 1st one:

![ss1](https://github.com/bootscore/bootscore/assets/26180119/f3265cb7-7145-4c57-9ce4-58748b25d223)

**Comment by crftwrk** [comment, marked as answer]
`bootscore-child-main/woocommerce/inc/wc-loop.php` isn't a template file and you cannot override it by storing a copy in your child. But I see, this function is still not pluggable and can't be overridden in child yet. 

Are you able to create a Pull Request and wrap this function in a `if` ... `endif` statement to make this function pluggable? This would be the fastest way to solve this. Otherwise, please open an issue that we have this on our tracking list. But it will take some time for us to fix this.

**Reply by xoneill007** [reply]
Thanks so much for the quick response!

If I can quickly wap an if-endif / test, I'll create a PR. Otherwise I'll report this as an issue.

You guys have create 'the best' WP Bootstrap - thanks for all you do. If the WebDEv side of my business picks up - I'm definitely contributing to this project!

**Reply by crftwrk** [reply]
You're welcome 👍

You can check any other function how the snippet is wrapped in `if` ... `endif`:

https://github.com/bootscore/bootscore/blob/7d7fbac9d0880c32625a864b863457e0d022d736/inc/loop.php#L15-L28

**Reply by xoneill007** [reply]
Thanks again, and sorry to pester...

Before I open a PR, just want to be sure I got this right:

1) In wc-loop.php (parent theme):

```
if (!function_exists('custom_loop_product_thumbnail')) :
   remove_action('woocommerce_before_shop_loop_item_title', 'woocommerce_template_loop_product_thumbnail', 10);
   add_action('woocommerce_before_shop_loop_item_title', 'custom_loop_product_thumbnail', 10);
   function custom_loop_product_thumbnail() {
     global $product;
     $size = 'woocommerce_thumbnail';
     $code = 'class=card-img-top';

     $image_size = apply_filters('single_product_archive_thumbnail_size', $size);

     echo $product ? $product->get_image($image_size, $code) : '';
   }
endif;
```

2) I copied wc-loop.php to the child theme (./woocommerce/inc/wc-loop.php) and added the following:

`remove_action('woocommerce_before_shop_loop_item_title', 'custom_loop_product_thumbnail', 10);`

3) I add the following to child's function.php:

```
function remove_thumbs() {
   require_once('woocommerce/inc/wc-loop.php');
}
add_action('woocommerce_before_shop_loop_item_title', 'remove_thumbs');
remove_action( 'woocommerce_before_shop_loop_item_title', 'woocommerce_template_loop_product_thumbnail', 10 );
```

However, I am still getting the same behavior - thumbs are removed from every product except the 1st (as depicted in the initial post / screenshot).

**Reply by crftwrk** [reply]
Thank you, almost...

1. Snippet in parent-theme is correct.
2. `wc-loop.php` is a function and not a template file. You cannot copy it to your child.
3. To override this function, simply copy code ***between*** `if` and `endif` to your child's `functions.php` as described here https://bootscore.me/documentation/functions/.
5. Now you can edit this function in your child.

Tested it and works fine. While we are here, let's change `custom_loop_product_thumbnail` to `bootscore_loop_product_thumbnail`. If you agree, please create a PR, we will merge it.

**Reply by xoneill007** [reply]
Thanks sir,

In the nth hour I finally figured that out :).

Still getting the strange thumbnail removal behavior...

This works (all thumbnails are removed):

```
// Remove product thumbnails from the shop loop:
remove_action( 'woocommerce_before_shop_loop_item_title', 'woocommerce_template_loop_product_thumbnail', 10 );

// Override bootScore function:
function bootscore_loop_product_thumbnail() {
   //remove_action( 'woocommerce_before_shop_loop_item_title', 'woocommerce_template_loop_product_thumbnail', 10 );
}
add_action('woocommerce_before_shop_loop_item_title', 'bootscore_loop_product_thumbnail');
```

This does not (all thumbnails are removed except for 1st product):

```
// Remove product thumbnails from the shop loop:
//remove_action( 'woocommerce_before_shop_loop_item_title', 'woocommerce_template_loop_product_thumbnail', 10 );

// Override bootScore function:
function bootscore_loop_product_thumbnail() {
   remove_action( 'woocommerce_before_shop_loop_item_title', 'woocommerce_template_loop_product_thumbnail', 10 );
}
add_action('woocommerce_before_shop_loop_item_title', 'bootscore_loop_product_thumbnail');
```

**Reply by crftwrk** [reply]
Ahh sorry, I see...

`remove_action` and `add_action` cannot be wrapped in a function. In this case, leave function as it is and do not try to override function in child. Override `content-product.php` by store a copy of it in your child and delete/comment hook line in template file: 

https://github.com/bootscore/bootscore/blob/7d7fbac9d0880c32625a864b863457e0d022d736/woocommerce/content-product.php#L45

If you just want to hide the product images in loop, you can hide them with a single line of CSS as well instead overriding the template file:

```css
.products .card-img-top {
  display: none;
}
```

### Edit

The entire function just adds the `card-img-top` class to product loop image, which is just the top border-radius. Since we started to reduce WC templates https://github.com/bootscore/bootscore/issues/333, more and more PHP was replaced by CSS. So for the card image in categories loop there is used a simple `@extend` instead filters:

https://github.com/bootscore/bootscore/blob/7d7fbac9d0880c32625a864b863457e0d022d736/scss/bootscore_woocommerce/_wc_loop.scss#L47-L51

However, I'm fine if we delete the entire function and replace snippet above with following snippet to fetch both, category and product loop images.

```scss
// Add .card-img-top to loop cards
.product-category .card,
.product.card {
  img {
    border-top-left-radius: $card-inner-border-radius;
    border-top-right-radius: $card-inner-border-radius;
  }
}
```

In this case, your code will work as expected:

```php
// Remove product thumbnails from the shop loop:
remove_action( 'woocommerce_before_shop_loop_item_title', 'woocommerce_template_loop_product_thumbnail', 10 );

```

Agree?

**Reply by xoneill007** [reply]
This makes sense, yes - thanks!

Apologies for the delay, had to set up an entirely new BSD machine + dev ENV.

I was about to open the PR and just noticed you edit.

So I'll del the function and modify?: bootscore/scss/bootscore_woocommerce/_wc_loop.scss

**Reply by crftwrk** [reply]
> So I'll del the function and modify?: bootscore/scss/bootscore_woocommerce/_wc_loop.scss

Exactly, this will do the trick.

**Reply by xoneill007** [reply]
All set:

https://github.com/bootscore/bootscore/pull/609

I've implemented on my local de site - works perfectly.

Apologies for dragging this out, thanks for the opportunity to contribute.
